basis-universal/src/transcoding/transcoder.rs

```diff
@@ -385,3 +385,90 @@ impl Drop for Transcoder {
         }
     }
 }
+
+pub struct LowLevelUastcTranscoder(*mut sys::LowLevelUastcTranscoder);
+
+impl Default for LowLevelUastcTranscoder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct SliceParametersUastc {
+    pub num_blocks_x: u32,
+    pub num_blocks_y: u32,
+    pub has_alpha: bool,
+    pub original_width: u32,
+    pub original_height: u32,
+}
+
+impl LowLevelUastcTranscoder {
+    /// Create a LowLevelUastcTranscoder
+    pub fn new() -> LowLevelUastcTranscoder {
+        transcoder_init();
+        unsafe { LowLevelUastcTranscoder(sys::low_level_uastc_transcoder_new()) }
+    }
+
+    pub fn transcode_slice(
+        &self,
+        data: &[u8],
+        slice_parameters: SliceParametersUastc,
+        decode_flags: DecodeFlags,
+        transcode_block_format: TranscoderBlockFormat,
+    ) -> Result<Vec<u8>, TranscodeError> {
+        let bc1_allow_threecolor_blocks = false;
+        let transcoder_state = std::ptr::null_mut();
+        let channel0 = 0;
+        let channel1 = 3;
+
+        let output_block_or_pixel_stride_in_bytes =
+            transcode_block_format.bytes_per_block_or_pixel();
+        let output_row_pitch_in_blocks_or_pixels =
+            (slice_parameters.original_width + transcode_block_format.block_width() - 1)
+                / transcode_block_format.block_width();
+        let output_rows_in_pixels = slice_parameters.original_height;
+
+        let output_size_bytes = (slice_parameters.num_blocks_x
+            * slice_parameters.num_blocks_y
+            * output_block_or_pixel_stride_in_bytes) as usize;
```

Comment:
should this be row_pitch * rows?

Reply:
I took a look at how this was being done for `TranscoderTextureFormat` and decided to copy over that code to `TranscoderBlockFormat`, expose underlying basis transcoder methods where available and implement replacements where they aren't, and then use that code for more consistent calculation logic. Hopefully this is preferable.